### PR TITLE
Add -o name for commands which use printer to output results

### LINKED ISSUE #5906: `kubectl get` etc. should support a `-o name` (or similar option) for showing only kind/name, and make mutations more verbose by default
With #5903 commands will support "resource/name" as an arg, which means having -q (show only type/name) becomes valuable for chaining:

    $ kubectl get pods -q | grep "-builder" | xargs kubectl delete

### PR DESCRIPTION
**Pull request by hurf**
Partially fixes: #5906

Some mutations (`label`, `run`, `expose`, `rolling-update`) are not covered by #10652. This patch will add '-o name' option for them. It also works for other command like `get` which uses resourcePrinter to output the results.

Added a new printer which prints 'resource/name' pair of a given
object.

### CODE REVIEW

**Comment by hurf** [comment]
This PR also works for #12638, I created a new PR for quicker review. @bgrant0607 @JanetKuo Can you take a look at this?

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl

**Comment by bgrant0607** [comment]
Thanks, @hurf. I would prefer these commands to be converted to call cmdutil.PrintSuccess, instead.

**Comment by bgrant0607** [comment]
For others: see also #12844

**Comment by hurf** [comment]
Sorry @bgrant0607 I didn't quit get you, do you mean I should also use `cmdutil.PrintSuccess` for `-o name` for all the commands? 

At first I did use `cmdutil.PrintSuccess`, but I found to do so I need to modify every command.
For commands like `create` which don't have `-o` flag before, it's fine and is done by #10652, but for commands like `label` and `get` which use `resourcePrinter` to print the results, the `-o` flag is already handled before printer is get. 
So I changed my idea to add a new Printer, then the commands don't need to be changed, they just call `printObj` function as they used to, but with `-o name`, a different printer is used.

**Comment by JanetKuo** [comment]
The idea was that instead of printing resources after each mutation, we should print messages like "pod frontend labeled."

So we don't need to print the results after applying mutations anymore. (Also, `get` isn't mutation.)

**Comment by hurf** [comment]
It seems we have a commond understanding @JanetKuo , in my understanding with `-o name` we're printing `resource/name` pairs, `pod fornted labeled` is printed when `-o` isn't specified, and `get` isn't a mutation, but there's requirement for it to print `resource/name` pairs, am I right?

I think this patch is intended to add `-o name` support for those commands which currently use resourcePrinter to print the result, nothing is done for the condition `-o` isn't specified. That #12844 's job.

**Comment by JanetKuo** [comment]
Based on earlier discussion (while I was writing #10652), `-o name` is used for mutations to simplify output message (being more machine-readable), and we don't need that for `get`. IMHO mutations don't need to print resources; therefore `-o name` is enough for them (other -o output formats are actually not necessary). 

@bgrant0607 Do we consider stop supporting -o yaml|template... in mutations?

**Comment by hurf** [comment]
I was thinking allow `get` output `resource/name` pairs for xarg to use. But if it's unnecessary and we want to drop `-o yaml|template...` for mutations, this patch would not help,  the output format change for both default and `-o name` can be done in  #12844.

**Comment by smarterclayton** [comment]
Today, in some commands `-o` actually means dry-run and is very useful.  In
others, it's less useful.  I think we want to consistently expose `-o yaml`
along with a clearer meaning of --dry-run for client tools.  That was the
concern before.  We should probably not be changing mutations in this issue
though for -o yaml / -o json

On Thu, Aug 20, 2015 at 2:55 AM, Victor HU <notifications@github.com> wrote:

> I was thinking allow get output resource/name pairs for xarg to use. But
> if it's unnecessary and we want to drop -o yaml|template... for
> mutations, this patch would not help, the output format change for both
> default and -o name can be done in #12844
> <https://github.com/kubernetes/kubernetes/pull/12844>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12845#issuecomment-132910675>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
-o name is useful for get -

    $ kubectl get pods -l with=aselector -o name | xargs delete pods

On Thu, Aug 20, 2015 at 10:42 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Today, in some commands `-o` actually means dry-run and is very useful.
> In others, it's less useful.  I think we want to consistently expose `-o
> yaml` along with a clearer meaning of --dry-run for client tools.  That was
> the concern before.  We should probably not be changing mutations in this
> issue though for -o yaml / -o json
>
> On Thu, Aug 20, 2015 at 2:55 AM, Victor HU <notifications@github.com>
> wrote:
>
>> I was thinking allow get output resource/name pairs for xarg to use. But
>> if it's unnecessary and we want to drop -o yaml|template... for
>> mutations, this patch would not help, the output format change for both
>> default and -o name can be done in #12844
>> <https://github.com/kubernetes/kubernetes/pull/12844>.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/12845#issuecomment-132910675>
>> .
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by hurf** [comment]
Then what should I do with this PR, I'm confused...

**Comment by JanetKuo** [comment]
If we're adding `-o name` to all commands (including `get`) that use printer, this change looks good to me. It'd be nice to have a test for `NamePrinter`, and it should be more clear in the help text on what output format `name` means (an example would help, too). 

Ping @bgrant0607 for additional comments - is it good to support `-o name` in addition to existing printing formats for mutations and `get`?

**Comment by JanetKuo** [comment]
Ref #11488 for `-o *` and `--dry-run`.

**Comment by hurf** [comment]
@JanetKuo - Added test for new Printer, and example for `get` command

File: pkg/api/meta/restmapper.go [hunk 1]
```diff
@@ -132,7 +132,7 @@ func (m *DefaultRESTMapper) Add(scope RESTScope, kind string, version string, mi
 }
 
 // kindToResource converts Kind to a resource name.
```
Comment (kargakis) [inline pkg/api/meta/restmapper.go]: uppercase the comment too

File: pkg/kubectl/resource_printer.go [hunk 2]
```diff
@@ -145,6 +149,56 @@ func (p *VersionedPrinter) PrintObj(obj runtime.Object, w io.Writer) error {
...
+				tpmeta := api.TypeMeta{
+					APIVersion: version,
+					Kind:       kind,
+				}
+				s := reflect.ValueOf(decodedObj).Elem()
+				s.FieldByName("TypeMeta").Set(reflect.ValueOf(tpmeta))
+				p.PrintObj(decodedObj, w)
+			}
+		} else {
+			return errors.New("The list object contains unrecognized items.")
```
Comment (kargakis) [inline pkg/kubectl/resource_printer.go]: lowercase the error string

**Comment by bgrant0607** [comment]
We should support `-o name` for all commands. Mutations without `-o name` should just print confirmation that the operation was performed in a consistent format, and should not print the resource details.

**Comment by bgrant0607** [comment]
`kubectl get ... -o name` should still contact the server, and should only print names of resources that exist. How to handle --dry-run, --local, etc. is being discussed here: https://github.com/kubernetes/kubernetes/pull/11368#issuecomment-133621152

**Comment by bgrant0607** [comment]
Do we support `-o yaml` and/or `-o json` for create/replace/patch today? If so, what does it print? Let's discuss dry run on #11368. If we support those output modes for mutations, I think they should dump the raw Status result by default, not the resource.

**Comment by smarterclayton** [comment]
Most mutations return objects today - including delete (because graceful
returns the pod at the moment we begin graceful deletion).  In our generate
code we use -o yaml to infer --dry-run.  I think for
create/update/replace/patch/delete it makes more sense for it to apply to
the output of the operation, and dry-run to be the correct way to indicate
not to contact the server as already discussed.

On Wed, Aug 26, 2015 at 12:19 AM, Brian Grant <notifications@github.com>
wrote:

> Do we support -o yaml and/or -o json for create/replace/patch today? If
> so, what does it print? Let's discuss dry run on #11368
> <https://github.com/kubernetes/kubernetes/pull/11368>. If we support
> those output modes for mutations, I think they should dump the raw Status
> result by default, not the resource.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12845#issuecomment-134807089>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by JanetKuo** [comment]
@hurf thanks! LGTM.

**Pull request merged**